### ISSUE TITLE
Hacks Required for 12 OH Links on CTP7

### ISSUE THREAD

**Issue by bdorney** [cms-gem-daq-project/gemctp7user#42]
## Brief summary of issue

For handling the 12 OH FW versions the following changes are necessary.

This is a "short-term" solution.

### Changes to `$ENV` variables:
Expect the following symlink to exist:

```bash
eagle61:/mnt/persistent/gemdaq$ ls -lhtr xml/*no_oh.pickle
lrwxrwxrwx    1 root     root          38 Feb  4 15:28 xml/gem_amc_top_no_oh.pickle -> gem_amc_top_AMCv3_7_1_NoOHBlock.pickle
```

This pickle file should be created from a modified version of the xml in which the optohybrid block has been removed.  To do this search for the commented text `OptoHybrid block` and remove the block that this corresponds too.

#### `gbt.py`
The `parseXML()` function needs to be told to use the slimmed down pickle file, i.e. change `programGBT(...)` to:

```python
def programGBT(ohSelect, gbtSelect, command):
    parseXML(pickleFileName="/mnt/persistent/gemdaq/xml/gem_amc_top_no_oh.pickle")

    initGbtRegAddrs()

    heading("Hello, I'm your GBT controller :)")
```

#### `rw_reg.py`
Need to change `parseXML()` function such that it can take an input pickle file name, i.e.:

```python
def parseXML(pickleFileName=None):                                                                                                                                                                                       
    if pickleFileName is None:
        print 'Open pickled address table if available ',ADDRESS_TABLE_TOP[:-3]+'pickle...'
        fname =  ADDRESS_TABLE_TOP[:-3] + "pickle"
    else:
        print 'Open pickled address table if available %s...'%(pickleFileName)
  fname = pickleFileName
```

## Context (for feature requests)

Presently issues exist because setting up a CTP7 relies on outdated `xhal` tag and with 12 OH's the pickle file exceeds the available RAM on the card.

**Comment by jsturdy** [comment]
@mexanick, can `parseXML` be changed to have an optional `skippedNodes` parameter?
This would then skip the mentioned nodes when looping over the tree, hopefully dynamically adjusting the parsing time.  Or does the parser already lose the time loading the XML into memory before it would be able to skip anything?

**Comment by mexanick** [comment]
we can not parse an XML (I mean human-readable XML) on the CTP7 due to absence of `lxml` (IIRC) which is required to parse nested XML files. So the parsing and packing into the `pickle` is happening at the host PC. In principle, we may try to add skipping of certain nodes during the parsing, but as far as I understand @evka85 can produce a stripped version of XML file without certain nodes. If this is the case, then I would request @evka85 to provide us a "short" XML version and stick with it.

**Comment by jsturdy** [comment]
OK, so the `parseXML` call here is simply loading the `pickle` file?
And that by definition will take the time that it takes to load the structure into memory and can't benefit from any dynamic exclusion of nodes?

Even if this is a "short term" "solution", creating a bunch of different address tables for specific purposes is a really djanky way to go, and having the ability to dynamically parse either XML or pickled address table is by far a more flexible and  overall a much better long term solution.

**Comment by mexanick** [comment]
> OK, so the `parseXML` call here is simply loading the `pickle` file?
And that by definition will take the time that it takes to load the structure into memory and can't benefit from any dynamic exclusion of nodes?

yes

> Even if this is a "short term" "solution", creating a bunch of different address tables for specific purposes is a really djanky way to go, and having the ability to dynamically parse either XML or pickled address table is by far a more flexible and overall a much better long term solution.

A long term solution is to use the corresponding RPC calls and LMDB, which is much faster in either way (and whose access time does not depend much on the size of the DB)

**Comment by jsturdy** [comment]
> A long term solution is to use the corresponding RPC calls and LMDB, which is much faster in either way (and whose access time does not depend much on the size of the DB)

Absolutely :-), however, I can still imagine that some generic, lightweight tools will still be developed during the testing phase (and maybe even the non GEM users of the `rw_reg` library) that would still benefit from streamlining the address table parsing (only if the request in the issue to modify the API is being considered, because otherwise, I agree, change nothing :-D )

**Comment by evka85** [comment]
Yeah the main problem is because of the generated nodes that are repeated many times, but perhaps we could store those in a bit smarter way so that we only keep one copy of the data in memory, and just apply an address offset when needed (e.g. for GEM_AMC.OH.OH1 node one could just point to GEM_AMC.OH.OH0 node and only store the address offset required to translate it to OH1). This way memory usage would be greatly reduced.
 
Also there are some other workarounds to reduce the pickle size e.g. don't store the documentation of each node, I have a feeling that documentation attributes actually take up most of that space, and from what I see, nobody is ever looking at it, so perhaps just storing documentation online would be sufficient.

**Comment by bdorney** [comment]
> I have a feeling that documentation attributes actually take up most of that space, and from what I see, nobody is ever looking at it, so perhaps just storing documentation online would be sufficient.

Perhaps it's my ignorance here, but `gem_reg.py/reg_interface.py` parses this pickle file and loads this information into memory.  And this documentation field is what is specifically returned when calling `doc <full node name>`.  Correct?

If this is true then it is definitely incorrect to say that "nobody is ever looking at it."  It's used extensively in the lab.

**Comment by evka85** [comment]
Yeah that's the one I had in mind. Ok, it's good to know that you guys are using it.
I just did some tests, and found that actual parsing the xml takes only 350ms on CTP7, and if I comment out the nested loop in make makeTree to go through the generated nodes then makeTree only takes 180ms on CTP7.  So if we change the data structure in a smart way, and instead of duplicating all data for generated nodes we only store some pointers to existing node (or something along those lines), I think we can speed it up and cut the memory consumption drastically, and get rid of the pickle hack..

I can look into it

**Comment by mexanick** [comment]
@evka85 by actual parsing you mean parsing the DOM without actually generating our "in-house" Node objects?

**Comment by evka85** [comment]
yeah
